lib/Ix/Processor/JMAP.pm

```diff
@@ -108,6 +109,65 @@ sub _sanity_check_calls ($self, $calls, $arg) {
   return;
 }
 
+sub expand_backrefs ($self, $ctx, $arg) {
+  my @backref_keys = map {; s/^#// ? $_ : () } keys %$arg;
+
+  my sub ref_error ($desc) {
+    Ix::Error::Generic->new({
+      error_type  => 'resultReference',
+      properties  => {
+        description => $desc,
+      },
+    });
+  }
+
+  return unless @backref_keys;
+
+  if (my @duplicated = grep {; exists $arg->{$_} } @backref_keys) {
+    return ref_error( "arguments present as both ResultReference and not: "
+                    .  join(q{, }, @duplicated));
+  }
+
+  my @sentences = $ctx->results_so_far->sentences;
+
+  for my $key (@backref_keys) {
+    my $ref  = delete $arg->{"#$key"};
+
+    unless ( _HASH0($ref)
+          && 3 == grep {; defined $ref->{$_} } qw(resultOf name path)
+    ) {
+      return ref_error("malformed ResultReference");
+    }
+
+    my ($sentence) = grep {; $_->client_id eq $ref->{resultOf} } @sentences;
+
+    unless ($sentence) {
+      return ref_error("no result for client id $ref->{resultOf}");
+    }
+
+    unless ($sentence->name eq $ref->{name}) {
+      return ref_error(
+        "first result for client id $ref->{resultOf} is not $ref->{name} but "
+        . $sentence->name,
+      );
+    }
+
+    my ($result, $error) = Ix::Util::resolve_modified_jpointer(
+      $ref->{path},
+      $sentence->arguments,
+    );
+
+    if ($error) {
+      return ref_error("error with path: $error");
+    }
+
+    # XXX: Is it safe to not clone $result? -- rjbs, 2018-01-10
+    $arg->{$key} = $result;
```

Comment:
I think there's two potential problems here.

First, not cloning the result - if something in the API modifies $result somehow, this could bust other calls down the line.

(Say we have three method calls, the latter 2 look at `#thing => first, resultSentence /args/1`, and the first of the the latter 2 calls modifies $result in the process of doing its work, the second call will get different data and/or fail.)

The other issue is one that already exists I guess: `$ctx->process_request` and `$ctx->handle_calls` can modify the incoming arguments which could be problematic for the caller's code...

Reply:
I've taken it as an unspoken rule that the args you pass to process_Request might be destroyed, and you need to accept that.  Internally, though, it's a problem.  If we assume that the hashref passed to one method gets destroyed, then passing the same hashref to another -- which is only possible with backrefs -- is not safe.  I'll add a dclone.